lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java

```diff
@@ -88,41 +90,115 @@ public void writeToken(
         dataStore.update(addressSessionItem);
     }
 
-    private void validateTokenRequest(Map<String, List<String>> queryParameters)
-            throws AccessTokenRequestException {
+    public TokenRequest createTokenRequest(String requestBody)
+            throws AccessTokenValidationException {
+        try {
+            URI arbitraryUri = URI.create("https://gds");
+            HTTPRequest request = new HTTPRequest(HTTPRequest.Method.POST, arbitraryUri);
+            request.setQuery(requestBody);
+            request.setContentType(ContentType.APPLICATION_URLENCODED.getType());
+
+            if (!request.getQueryParameters()
+                    .keySet()
+                    .containsAll(
+                            Set.of(CODE, CLIENT_ID, CLIENT_ASSERTION_TYPE, CLIENT_ASSERTION))) {
+                throw new AccessTokenValidationException("missing request parameters");
+            }
+
+            return TokenRequest.parse(request);
+        } catch (com.nimbusds.oauth2.sdk.ParseException e) {
+            throw new AccessTokenValidationException(e);
+        }
+    }
 
-        var listHelper = new ListUtil();
+    public TokenRequest validateTokenRequest(TokenRequest tokenRequest)
+            throws AccessTokenValidationException {
+        try {
 
-        var authorizationCode =
-                listHelper.getValueOrThrow(
-                        queryParameters.getOrDefault(CODE, Collections.emptyList()));
-        var redirectUri =
-                listHelper.getValueOrThrow(
-                        queryParameters.getOrDefault(REDIRECT_URI, Collections.emptyList()));
-        var grantType =
-                listHelper.getValueOrThrow(
-                        queryParameters.getOrDefault(GRANT_TYPE, Collections.emptyList()));
+            ClientAuthentication clientAuthentication = tokenRequest.getClientAuthentication();
+            PrivateKeyJWT privateKeyJWT = (PrivateKeyJWT) clientAuthentication;
```

Comment:
@amanm-alphagov this is the important bit of the PR - where we get a PrivateKeyJWT, then verify its payload and signing